### ISSUE TITLE
ec2_group fails to add/delete rules

### ISSUE THREAD

**Issue by alexclifford** [ansible/ansible-modules-core#895]
##### Issue Type:

Bug Report
##### Ansible Version:

Ansible 1.8.4
##### Environment:

Ubuntu 14.04 LTS
##### Summary:

When running an ec2_group module Playbook for the same VPC/Security Group but with different defined rules the results are not consistent, and seemingly miss (randomly) adding all defined rules.
##### Steps To Reproduce:

```
archer:alex-21:22:28 {master} ~/work/gitlab/ansible-playbooks$ cat plays/test1.yml 

---
- hosts: localhost
  connection: local
  gather_facts: False

  tasks:
  - name: create ec2 group
    local_action:
      module: ec2_group
      region: "ap-southeast-2"
      name: "testing1234"
      description: "testing1234"
      vpc_id: "vpc-b5f631d0"
      rules:
      - proto: tcp
        from_port: 22
        to_port: 22
        cidr_ip: "1.1.1.1/32"
      - proto: tcp
        from_port: 22
        to_port: 22
        cidr_ip: "2.2.2.2/32"
      - proto: tcp
        from_port: 22
        to_port: 22
        cidr_ip: "3.3.3.3/32"
      - proto: tcp
        from_port: 22
        to_port: 22
        cidr_ip: "4.4.4.4/32"
      - proto: tcp
        from_port: 80
        to_port: 80
        cidr_ip: "0.0.0.0/0"
      - proto: tcp
        from_port: 443
        to_port: 443
        cidr_ip: "5.5.5.0/24"
    register: ec2_group

  - debug: var=ec2_group

archer:alex-21:22:31 {master} ~/work/gitlab/ansible-playbooks$ cat plays/test2.yml

---
- hosts: localhost
  connection: local
  gather_facts: False

  tasks:
  - name: create ec2 group
    local_action:
      module: ec2_group
      region: "ap-southeast-2"
      name: "testing1234"
      description: "testing1234"
      vpc_id: "vpc-b5f631d0"
      rules:
      - proto: tcp
        from_port: 22
        to_port: 22
        cidr_ip: "2.2.2.2/32"
      - proto: tcp
        from_port: 22
        to_port: 22
        cidr_ip: "4.4.4.4/32"
      - proto: tcp
        from_port: 80
        to_port: 80
        cidr_ip: "0.0.0.0/0"
      - proto: tcp
        from_port: 443
        to_port: 443
        cidr_ip: "5.5.5.5/32"
    register: ec2_group

  - debug: var=ec2_group

archer:alex-21:21:17 {master} ~/work/gitlab/ansible-playbooks$ time ansible-playbook -vvvv plays/test1.yml 
PLAY [localhost] ************************************************************** 

TASK: [create ec2 group] ****************************************************** 
<127.0.0.1> REMOTE_MODULE ec2_group description=testing1234 name=testing1234 region=ap-southeast-2 vpc_id=vpc-b5f631d0
<127.0.0.1> EXEC ['/bin/sh', '-c', 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1423736477.63-15178099669983 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1423736477.63-15178099669983 && echo $HOME/.ansible/tmp/ansible-tmp-1423736477.63-15178099669983']
<127.0.0.1> PUT /tmp/tmpk_fmxX TO /home/alex/.ansible/tmp/ansible-tmp-1423736477.63-15178099669983/ec2_group
<127.0.0.1> EXEC ['/bin/sh', '-c', u'LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python /home/alex/.ansible/tmp/ansible-tmp-1423736477.63-15178099669983/ec2_group; rm -rf /home/alex/.ansible/tmp/ansible-tmp-1423736477.63-15178099669983/ >/dev/null 2>&1']
changed: [localhost -> 127.0.0.1] => {"changed": true, "group_id": "sg-44ea4221"}

TASK: [debug var=ec2_group] *************************************************** 
ok: [localhost] => {
    "ec2_group": {
        "changed": true, 
        "group_id": "sg-44ea4221", 
        "invocation": {
            "module_args": "", 
            "module_name": "ec2_group"
        }
    }
}

PLAY RECAP ******************************************************************** 
localhost                  : ok=2    changed=1    unreachable=0    failed=0   


real    0m3.975s
user    0m0.318s
sys    0m0.060s

archer:alex-21:21:21 {master} ~/work/gitlab/ansible-playbooks$ aws ec2 describe-security-groups --group-ids sg-44ea4221
{
    "SecurityGroups": [
        {
            "IpPermissionsEgress": [
                {
                    "IpProtocol": "-1", 
                    "IpRanges": [
                        {
                            "CidrIp": "0.0.0.0/0"
                        }
                    ], 
                    "UserIdGroupPairs": []
                }
            ], 
            "Description": "testing1234", 
            "IpPermissions": [
                {
                    "ToPort": 22, 
                    "IpProtocol": "tcp", 
                    "IpRanges": [
                        {
                            "CidrIp": "1.1.1.1/32"
                        }, 
                        {
                            "CidrIp": "2.2.2.2/32"
                        }, 
                        {
                            "CidrIp": "3.3.3.3/32"
                        }, 
                        {
                            "CidrIp": "4.4.4.4/32"
                        }
                    ], 
                    "UserIdGroupPairs": [], 
                    "FromPort": 22
                }, 
                {
                    "ToPort": 80, 
                    "IpProtocol": "tcp", 
                    "IpRanges": [
                        {
                            "CidrIp": "0.0.0.0/0"
                        }
                    ], 
                    "UserIdGroupPairs": [], 
                    "FromPort": 80
                }, 
                {
                    "ToPort": 443, 
                    "IpProtocol": "tcp", 
                    "IpRanges": [
                        {
                            "CidrIp": "5.5.5.0/24"
                        }
                    ], 
                    "UserIdGroupPairs": [], 
                    "FromPort": 443
                }
            ], 
            "GroupName": "testing1234", 
            "VpcId": "vpc-b5f631d0", 
            "OwnerId": "014746447491", 
            "GroupId": "sg-44ea4221"
        }
    ]
}

archer:alex-21:21:22 {master} ~/work/gitlab/ansible-playbooks$ time ansible-playbook -vvvv plays/test2.yml 
PLAY [localhost] ************************************************************** 

TASK: [create ec2 group] ****************************************************** 
<127.0.0.1> REMOTE_MODULE ec2_group description=testing1234 name=testing1234 region=ap-southeast-2 vpc_id=vpc-b5f631d0
<127.0.0.1> EXEC ['/bin/sh', '-c', 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1423736482.65-174316560142584 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1423736482.65-174316560142584 && echo $HOME/.ansible/tmp/ansible-tmp-1423736482.65-174316560142584']
<127.0.0.1> PUT /tmp/tmpJFGKqv TO /home/alex/.ansible/tmp/ansible-tmp-1423736482.65-174316560142584/ec2_group
<127.0.0.1> EXEC ['/bin/sh', '-c', u'LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python /home/alex/.ansible/tmp/ansible-tmp-1423736482.65-174316560142584/ec2_group; rm -rf /home/alex/.ansible/tmp/ansible-tmp-1423736482.65-174316560142584/ >/dev/null 2>&1']
changed: [localhost -> 127.0.0.1] => {"changed": true, "group_id": "sg-44ea4221"}

TASK: [debug var=ec2_group] *************************************************** 
ok: [localhost] => {
    "ec2_group": {
        "changed": true, 
        "group_id": "sg-44ea4221", 
        "invocation": {
            "module_args": "", 
            "module_name": "ec2_group"
        }
    }
}

PLAY RECAP ******************************************************************** 
localhost                  : ok=2    changed=1    unreachable=0    failed=0   


real    0m2.858s
user    0m0.334s
sys    0m0.055s

archer:alex-21:21:25 {master} ~/work/gitlab/ansible-playbooks$ aws ec2 describe-security-groups --group-ids sg-44ea4221
{
    "SecurityGroups": [
        {
            "IpPermissionsEgress": [
                {
                    "IpProtocol": "-1", 
                    "IpRanges": [
                        {
                            "CidrIp": "0.0.0.0/0"
                        }
                    ], 
                    "UserIdGroupPairs": []
                }
            ], 
            "Description": "testing1234", 
            "IpPermissions": [
                {
                    "ToPort": 22, 
                    "IpProtocol": "tcp", 
                    "IpRanges": [
                        {
                            "CidrIp": "4.4.4.4/32"
                        }
                    ], 
                    "UserIdGroupPairs": [], 
                    "FromPort": 22
                }, 
                {
                    "ToPort": 80, 
                    "IpProtocol": "tcp", 
                    "IpRanges": [
                        {
                            "CidrIp": "0.0.0.0/0"
                        }
                    ], 
                    "UserIdGroupPairs": [], 
                    "FromPort": 80
                }, 
                {
                    "ToPort": 443, 
                    "IpProtocol": "tcp", 
                    "IpRanges": [
                        {
                            "CidrIp": "5.5.5.5/32"
                        }
                    ], 
                    "UserIdGroupPairs": [], 
                    "FromPort": 443
                }
            ], 
            "GroupName": "testing1234", 
            "VpcId": "vpc-b5f631d0", 
            "OwnerId": "014746447491", 
            "GroupId": "sg-44ea4221"
        }
    ]
}

archer:alex-21:21:36 {master} ~/work/gitlab/ansible-playbooks$ time ansible-playbook -vvvv plays/test2.yml
PLAY [localhost] ************************************************************** 

TASK: [create ec2 group] ****************************************************** 
<127.0.0.1> REMOTE_MODULE ec2_group description=testing1234 name=testing1234 region=ap-southeast-2 vpc_id=vpc-b5f631d0
<127.0.0.1> EXEC ['/bin/sh', '-c', 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1423736509.2-214391272127452 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1423736509.2-214391272127452 && echo $HOME/.ansible/tmp/ansible-tmp-1423736509.2-214391272127452']
<127.0.0.1> PUT /tmp/tmphbzwbf TO /home/alex/.ansible/tmp/ansible-tmp-1423736509.2-214391272127452/ec2_group
<127.0.0.1> EXEC ['/bin/sh', '-c', u'LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python /home/alex/.ansible/tmp/ansible-tmp-1423736509.2-214391272127452/ec2_group; rm -rf /home/alex/.ansible/tmp/ansible-tmp-1423736509.2-214391272127452/ >/dev/null 2>&1']
changed: [localhost -> 127.0.0.1] => {"changed": true, "group_id": "sg-44ea4221"}

TASK: [debug var=ec2_group] *************************************************** 
ok: [localhost] => {
    "ec2_group": {
        "changed": true, 
        "group_id": "sg-44ea4221", 
        "invocation": {
            "module_args": "", 
            "module_name": "ec2_group"
        }
    }
}

PLAY RECAP ******************************************************************** 
localhost                  : ok=2    changed=1    unreachable=0    failed=0   


real    0m2.137s
user    0m0.305s
sys    0m0.077s

archer:alex-21:21:51 {master} ~/work/gitlab/ansible-playbooks$ aws ec2 describe-security-groups --group-ids sg-44ea4221
{
    "SecurityGroups": [
        {
            "IpPermissionsEgress": [
                {
                    "IpProtocol": "-1", 
                    "IpRanges": [
                        {
                            "CidrIp": "0.0.0.0/0"
                        }
                    ], 
                    "UserIdGroupPairs": []
                }
            ], 
            "Description": "testing1234", 
            "IpPermissions": [
                {
                    "ToPort": 22, 
                    "IpProtocol": "tcp", 
                    "IpRanges": [
                        {
                            "CidrIp": "4.4.4.4/32"
                        }, 
                        {
                            "CidrIp": "2.2.2.2/32"
                        }
                    ], 
                    "UserIdGroupPairs": [], 
                    "FromPort": 22
                }, 
                {
                    "ToPort": 80, 
                    "IpProtocol": "tcp", 
                    "IpRanges": [
                        {
                            "CidrIp": "0.0.0.0/0"
                        }
                    ], 
                    "UserIdGroupPairs": [], 
                    "FromPort": 80
                }, 
                {
                    "ToPort": 443, 
                    "IpProtocol": "tcp", 
                    "IpRanges": [
                        {
                            "CidrIp": "5.5.5.5/32"
                        }
                    ], 
                    "UserIdGroupPairs": [], 
                    "FromPort": 443
                }
            ], 
            "GroupName": "testing1234", 
            "VpcId": "vpc-b5f631d0", 
            "OwnerId": "014746447491", 
            "GroupId": "sg-44ea4221"
        }
    ]
}
```
##### Expected Results:

The first run of plays/test2.yml should change the group correctly (not requiring a second run).
##### Actual Results:

After plays/test2.yml is run and checked one of the port 22 rules is just not there, but reappears after running the plays/test2.yml again. If I delete the Security Group and start the process again I get the same weird results. ec2_group just gets confused and doesn't add all the defined rules.

**Comment by alexclifford** [comment]
This is a still a big problem that happens a lot. The rules array defined still seems to randomly not remove or add all the specified rules. I can't trust running this on production hosts.

I tested a few more times and on 1.9.0.1 I might be having better results - was any code surrounding this functionality fixed since 1.8.4?

**Comment by alexclifford** [comment]
It looks like something in a later changelog mentions something along the lines of this behaviour but I don't know the exact issue # so I can't link to it. Either way this can be closed as completed.
